sw/device/lib/crypto/impl/mac.c

```diff
@@ -9,23 +9,202 @@
+                              hmac_mode_t *hmac_mode, size_t *block_size) {
+  switch (key_mode) {
+    case kOtcryptoKeyModeHmacSha256:
+      *hmac_mode = kHmacModeHmac256;
+      *block_size = kHmacSha256BlockWords;
+      break;
+    case kOtcryptoKeyModeHmacSha384:
+      *hmac_mode = kHmacModeHmac384;
+      // Note that HMAC-384 and HMAC-512 have the same internal block size.
+      *block_size = kHmacSha512BlockWords;
+      break;
+    case kOtcryptoKeyModeHmacSha512:
+      *hmac_mode = kHmacModeHmac512;
+      *block_size = kHmacSha512BlockWords;
+      break;
+    default:
+      return OTCRYPTO_BAD_ARGS;
+  }
+  return OTCRYPTO_OK;
+}
+
+/**
+ * Pad or hash HMAC key to full block. I.e. compute K0 as defined in FIPS 198-1,
+ * Section 4, Steps 1-3.
+ *
+ * For HMAC-256, `processed_key` is 512 bits, and for HMAC-384/512,
+ * `processed_key` is 1024 bits.
+ *
+ * The caller must allocate `processed_key` buffer with internal block size.
+ *
+ * @param key The blinded input key.
+ * @param[out] processed_key Padding/hashed key whose size matches the internal
+ * block size of the hash function.
+ * @return Result of the operation.
+ */
+static status_t hmac_key_process(const otcrypto_blinded_key_t *key,
+                                 uint32_t *processed_key) {
+  otcrypto_hash_mode_t hash_mode;
+  size_t block_size;
+  size_t digest_size;
+  switch (key->config.key_mode) {
+    case kOtcryptoKeyModeHmacSha256:
+      hash_mode = kOtcryptoHashModeSha256;
+      block_size = kHmacSha256BlockWords;
+      digest_size = kHmacSha256DigestWords;
+      break;
+    case kOtcryptoKeyModeHmacSha384:
+      hash_mode = kOtcryptoHashModeSha384;
+      // Note that HMAC-384 and HMAC-512 have the same internal block size.
+      block_size = kHmacSha512BlockWords;
+      digest_size = kHmacSha384DigestWords;
+      break;
+    case kOtcryptoKeyModeHmacSha512:
+      hash_mode = kOtcryptoHashModeSha512;
+      block_size = kHmacSha512BlockWords;
+      digest_size = kHmacSha512DigestWords;
+      break;
+    default:
+      return OTCRYPTO_BAD_ARGS;
+  }
+
+  // HMAC HWIP does not support masking, so we need to unmask the key.
+  size_t unmasked_key_len = keyblob_share_num_words(key->config);
+  uint32_t unmasked_key[unmasked_key_len];
+  HARDENED_TRY(keyblob_key_unmask(key, unmasked_key_len, unmasked_key));
+
+  // Pre-populate with 0s, in order to pad keys smaller than the internal
+  // block size, according to FIPS 198-1, Section 4.
+  memset(processed_key, 0, block_size * sizeof(uint32_t));
+  // If the key is larger than the internal block size, we need to hash it
+  // according to FIPS 198-1, Section 4, Step 2.
+  if (key->config.key_length > block_size * sizeof(uint32_t)) {
+    otcrypto_hash_digest_t key_digest = {
+        .mode = hash_mode,
+        .data = processed_key,
+        .len = digest_size,
+    };
+    otcrypto_const_byte_buf_t msg_buf = {
+        .len = key->config.key_length,
+        .data = (uint8_t *)unmasked_key,
```

Comment:
nit: technically this violates strict aliasing, but `unsigned char *` is legal and equivalent:
```suggestion
        .data = (unsigned char *)unmasked_key,
```

Reply:
My understanding on the strict aliasing rules is limited, so I will go with your suggestion here. Thanks.